### PR TITLE
RooFit::MultiProcess & TestStatistics part 4: RooMinimizerFcn refactoring

### PR DESCRIPTION
**Pull request by egpbos**
This PR refactors RooMinimizerFcn into two separate classes: RooAbsMinimizerFcn and a new RooMinimizerFcn derived from that. RooAbsMinimizerFcn contains all common synchronization "logistics" necessary between RooFit and Minuit function parameters. The implementations of this abstract base class are then expected to supply the actual function evaluation parts. This allows us to later implement classes that calculate the gradient outside of Minuit. We will provide both a serial implementation of this and later on a parallel one using the MultiProcess framework.

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

### CODE REVIEW

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx
Comment (guitargeek) [inline roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:15]: Is this `ifndef` really needed? When would we build with `__ROOFIT_NOROOMINIMIZER`?

File: roofit/roofitcore/src/RooMinimizer.cxx
Comment (guitargeek) [inline roofit/roofitcore/src/RooMinimizer.cxx:962]: I would not outline these one-liners. They should better be inlined in the header file. There are also many such one-liners in RooAbsMininmizerFcn, but since they are copy pasted from the original file, I'm fine with keeping them there. But for new code, it would be nice to inline.

File: roofit/roofitcore/src/RooMinimizerFcn.cxx
Comment (guitargeek) [inline roofit/roofitcore/src/RooMinimizerFcn.cxx:152]: These two functions `getFunctionName` and `getFunctionTitle` can also be inlined in the header to allow for better compiler optimization and improve code readability.

File: roofit/roofitcore/inc/RooMinimizerFcn.h
Comment (guitargeek) [inline roofit/roofitcore/inc/RooMinimizerFcn.h:46]: `get_nDim` is not RooFit variable naming convention with the underscore in the name. `getNDim` would be more consistent.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx
Comment (egpbos) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:15]: It was already there, but seems unused. Wouter confirms that it is obsolete, so could be removed in all places that it is still used. Separate PR probably?

Comment (guitargeek) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:15]: Ah sorry, I didn't realize it was there before. I agree, everything that is cleaning up old stuff is a task for a separate PR.

File: roofit/roofitcore/inc/RooMinimizerFcn.h
Comment (hageboeck) [inline roofit/roofitcore/inc/RooMinimizerFcn.h:38]: Multiple inheritance from classes that are not abstract isn't nice in C++. Now all the stuff such as having two separate data sections, pointer changes on casting etc are going to happen.
If possible, make `AbsMinimizerFcn` really abstract, i.e. **no** members.

File: roofit/roofitcore/inc/RooMinimizerFcn.h
Comment (hageboeck) [inline roofit/roofitcore/inc/RooMinimizerFcn.h:36]: Those forward declarations are on honourable attempt, but since the abstract version includes all those things, they are unnecessary.
To detect include errors in other classes, the best ordering is
```c++
#include "MyBaseClass.h"

#include "RooFitStuff.h"
#include "RooFitStuff2.h"

#include "RootStuff.h"

#include <stdlib>
```

But again: Forward declare if possible.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx
Comment (hageboeck) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:15]: No, not needed. Agree about a separate PR, where things like [this part here](https://github.com/root-project/root/blob/7c4f9321a3fcbda598da6b815409a9a5ed72a53e/roofit/roofitcore/src/RooAbsPdf.cxx#L1599-L1783) could be condensed down into the one branch that's actually used.

File: roofit/roofitcore/inc/RooAbsMinimizerFcn.h
Comment (hageboeck) [inline roofit/roofitcore/inc/RooAbsMinimizerFcn.h:45]: Naming conventions? Add doxygen documentation, please.

File: roofit/roofitcore/inc/RooAbsMinimizerFcn.h
Comment (hageboeck) [inline roofit/roofitcore/inc/RooAbsMinimizerFcn.h:54]: Doxygen also here.

File: roofit/roofitcore/inc/RooAbsMinimizerFcn.h
Comment (hageboeck) [inline roofit/roofitcore/inc/RooAbsMinimizerFcn.h:57]: Doxygen.

File: roofit/roofitcore/inc/RooAbsMinimizerFcn.h
Comment (hageboeck) [inline roofit/roofitcore/inc/RooAbsMinimizerFcn.h:102]: Doxygen also here. This is especially important since this class defines the interface, so users (implementers) should be able to look up what the functions are supposed to do.

File: roofit/roofitcore/inc/RooAbsMinimizerFcn.h
Comment (hageboeck) [inline roofit/roofitcore/inc/RooAbsMinimizerFcn.h:127]: Try not to have data members in abstract classes.
If there's no way around them, better make them private. Making them protected increases the amount of code that has access to them, and which will make refactoring harder, since there's more code that has to be touched.

File: roofit/roofitcore/inc/RooAbsMinimizerFcn.h
Comment (hageboeck) [inline roofit/roofitcore/inc/RooAbsMinimizerFcn.h:123]: When touching this anyway: `unique_ptr`?

File: roofit/roofitcore/src/RooMinimizerFcn.cxx
Comment (hageboeck) [reply roofit/roofitcore/src/RooMinimizerFcn.cxx:152]: Maybe, maybe not. I left some comments to bring down the includes in the headers. If you follow that route, it's actually better to forward declare `RooAbsReal`, and have those functions in the `cxx`.
If `RooAbsReal` needs to be included in the header, I would suggest inlining.

**Review by hageboeck — COMMENTED** [review]
I have some suggestions regarding keeping headers small and with few dependencies to speed up compilation.
I also think that for an abstract class (an interface), `RooAbsMinimizer` has too many members and too many includes. Consider those suggestions optional, but if this code is touched, it could as well be modernised if time permits.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx
Comment (egpbos) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:15]: I will just remove them when I encounter them.

File: roofit/roofitcore/src/RooMinimizerFcn.cxx
Comment (guitargeek) [reply roofit/roofitcore/src/RooMinimizerFcn.cxx:152]: Good point, if inlining means to add mode includes that would otherwise not be necessary it should not be done.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx
Comment (egpbos) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:15]: Oh weird, I didn't see your replies when I typed that I would just remove them, should have refreshed the page. I didn't actually notice the part that you linked @hageboeck. The ifdef's I encountered so far were completely inconsequential (those that switch on/off the entire file, which I think really is useless), so perhaps those can be safely removed while the one you mention would need a little extra motivation and double checking.

**Comment by egpbos** [comment]
@hageboeck Thanks for the extensive review! I will implement some changes based on that and @guitargeek's comments today.

In the meantime, perhaps it would be useful to check out #8596 as well. That PR introduces the second concrete MinimizerFcn. Possibly, seeing `RooAbsMinimizerFcn` "in action" through that new class (`RooGradMinimizerFcn`) will shed a different light on the design choices for this abstract class. Basically, we previously just had two classes, `RooMinimizerFcn` and `RooGradMinimizerFcn` (and then later on **another** one, which will be introduced in another PR this week) with tons of code duplication. `RooAbsMinimizerFcn` simply factors that stuff out into one class.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx
Comment (hageboeck) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:15]: Yes, please do remove them here, and Jonas can take care of the ones that are left in roofitcore.
#8604

**Comment by guitargeek** [comment]
Hi @egpbos, FYI, there are conflicts in this PR now that https://github.com/root-project/root/pull/8609 got merged.

**Comment by egpbos** [comment]
@guitargeek thanks for the heads-up, I will fix those later.

I have made a new commit, which addresses your and @hageboeck's review comments as much as possible. Point by point, let me quickly explain what changed:

- "Upgraded" comments on new functionality to doxygen. I did not doxygen document everything, because most functionality was copy-pasted from the old RooMinimizerFcn. The new things are now documented though.
- Renamed snake_case to dromedary.
- Use unique_ptr for parameter lists, as suggested by Stephan.
- Remove a few obsolete __ROOFIT_NOROOMINIMIZER flags (I guess these are part of the merge conflicts).
- I realized that part of what I changed in PR #8596 could (and imho should) have been introduced here already. I moved functionality to RooAbsMinimizerFcn, namely storage of the to-be-minimized function and constant term optimization information, but did not yet remove it (in this PR) from RooMinimizer. Now it has been removed.
- While going over the above, I realized I had actually *completely* forgot to add back in a backwards compatible way of enabling offsetting for numerical precision! This has now been correctly added back in.

Then, about what I did not change:

- I did not have time for doing the include reordering/optimization that @hageboeck suggested. I tried to quickly figure it out, but I couldn't change things fast enough and still be able to guarantee that everything will still work in the later PRs that build on this one. I agree it would be nice, but I hope we can postpone this to some future PR, perhaps after the other additions have been merged.
- As mentioned earlier, RooAbsMinimizerFcn is basically a way to reduce code duplication. This should be seen in light of RooGradMinimizerFcn as well (see again PR #8596) and will also be used for a later addition. I think removing all data members and also the protected vs private suggestion would complicate this goal.

I think it's good to merge like this (after fixing the conflict of course), but let me know if you have further suggestions!

**Comment by egpbos** [comment]
Fixed the merge conflict as well now.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx
Comment (guitargeek) [inline roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:33]:
```suggestion
#include "TClass.h"
```

Because of the `arg->IsA()->InheritsFrom()` in the same file below. Unfortunately ROOT doesn't build on my system before this change.

File: roofit/roofitcore/inc/RooMinimizer.h
Comment (guitargeek) [inline roofit/roofitcore/inc/RooMinimizer.h:112]: I only catch this now, but if you remove or add non-transient member variables to a ROOT class you have to update the class version. Which means that at the end of this file, you have update this:
```C++
ClassDef(RooMinimizer,0) // RooFit interface to ROOT::Fit::Fitter
```
to this
```
ClassDef(RooMinimizer,0) // RooFit interface to ROOT::Fit::Fitter
```

**Review by guitargeek — CHANGES_REQUESTED** [review]
Hi Patrick, I have two final requests: one to make the code build also on my system, and one related to the class version of `RooMinimizer` changing a ROOT class and forgetting to update the class version can have bad consequences for users, so it would really be better to do this already in this PR.

By the way, once the class version has been updated once in a given release development cycle, it doesn't need to be updated again in the same development cycle.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx
Comment (egpbos) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:33]: Weird how those includes work differently on different systems sometimes... My IDE (CLion on macOS) even immediately notes that the include is unused. Perhaps something to do with platform-based compile flags then? Anyway, fixed!

File: roofit/roofitcore/inc/RooMinimizer.h
Comment (egpbos) [reply roofit/roofitcore/inc/RooMinimizer.h:112]: Ah good to know about the release cycle point, I indeed intended to bump the number in the final edits. Will do now instead.

**Comment by egpbos** [comment]
Ok, I made the two requested changes @guitargeek, thanks for the sharp reviews and support!

File: roofit/roofitcore/inc/RooMinimizer.h
Comment (guitargeek) [reply roofit/roofitcore/inc/RooMinimizer.h:112]: I see, good that you are keeping track of the class versions :+1: I just prefer to do at at the first edit in the release cycle so we won't forget.

**Review by guitargeek — APPROVED** [review]
Looks good to me!
This PR splits up the RooMinimzerFunc into a RooAbsMinimizerFunc and an actual RooMinimizerFunc that inherits from it. This is to prepare for further RooFit developments. Most review comments have been addressed, for more details see the summary in https://github.com/root-project/root/pull/8569#issuecomment-875559598.

**Comment by hageboeck** [comment]
@guitargeek @egpbos

Stop, stop!! Needs e3b73a8 to be reverted before releasing ROOT!

`ClassDef(..., 0)` means: Don't support I/O.
Now, you switched on I/O capabilities, so please revert! When people start writing minimiser states to files, you will have to support deserialisation in the future. It's better to not support that.

**That by the way also holds for any classes the are introduced in the MP-PRs. Please consider well wether they are allowed to support I/O or not.**

**Comment by hageboeck** [comment]
@guitargeek If the class *had* supported I/O before, the situation would be even worse: In general
- You would need to write an instance of the class to a file, and save this file somewhere in the test suite.
- You would need to write a schema evolution rule to convert old data members to new ones (not in this case, since removing).
- You would need to add a test that reads back the old class instance, and tests that the new (in-memory) version is in a consistent state.

No harm done here, since you can just revert the one line, but please double check that none of the other PRs destroy I/O capabilities of RooFit.

**Comment by egpbos** [comment]
Very sharp, thanks @hageboeck, I had no idea. Shall I do a revert PR for that one line?

**Comment by egpbos** [comment]
Ah great, you already did, thanks!

**Comment by egpbos** [comment]
Ah no, haha, it's an issue, ok, so shall I just make a PR to make the change or do you want to do it yourself? Is there some git procedure I should observe? I think `git revert` in this case is more work than just changing the one line.

**Comment by hageboeck** [comment]
I just created an issue from the comment, so yes, please add a PR! 👍

**Comment by guitargeek** [comment]
Ouch, that's painful! I totally fell for this, totally my fault. I suggested to increase the class number in the other PR. Now I will always use this as an example for why magic numbers are dangerous :D

Don't worry about this, @egpbos, I'll make a PR to revert the version number increase myself. I'll also add a comment to all RooFit classes with `ClassDef(..., 0)` so this mistake doesn't happen again.

**Comment by egpbos** [comment]
Ah, damn, I didn't read your comment in time @guitargeek, I already made a PR as well. Please feel free to close. And I agree, magic numbers bad :D

**Comment by egpbos** [comment]
Btw, if no I/O is necessary, is the ClassDef itself actually necessary?

Alternatively (if it is necessary): what about a macro overload without version number argument? And a static check in ClassDef that fails compilation on zero, referring to the new versionless one instead.

**Comment by hageboeck** [comment]
It's necessary if you want ROOT's type inspection capabilities to work, e.g.
`this->IsA()->InheritsFrom(SomeThing::Class())`

In most cases, this doesn't have advantages over `dynamic_cast`, so I guess that you *mostly* want the ClassDef added when I/O is needed.

**Comment by guitargeek** [comment]
Thanks Patrick for the PR!

**Pull request merged**
